include/libtorrent/kademlia/rpc_manager.hpp

```diff
@@ -94,15 +94,29 @@ class TORRENT_EXTRA_EXPORT rpc_manager
 	void check_invariant() const;
 #endif
 
-	void* allocate_observer();
-	void free_observer(void* ptr);
+	template <typename T, typename... Args>
+	std::shared_ptr<T> allocate_observer(Args&&... args)
+	{
+		void* ptr = allocate_observer();
+		if (ptr == nullptr) return std::shared_ptr<T>();
+
+		auto deleter = [this](observer* o)
+		{
+			o->~observer();
+			free_observer(o);
+		};
+		return std::shared_ptr<T>(new (ptr) T(args...), deleter);
```

Comment:
please use T(std::forward<Args>(args)...) (I think that's how you'd type it at least)